model/test_results.go

```diff
@@ -181,79 +183,112 @@ func (t *TestResults) Append(ctx context.Context, results []TestResult) error {
 
 	}
 
-	return errors.Wrap(t.appendTestNames(ctx, testNames), "problem updating log metadata during upload")
+	return nil
 }
 
-func (t *TestResults) appendTestNames(ctx context.Context, testNames []string) error {
+// Download returns a TestResult slice with the corresponding results stored in
+// the offline blob storage. The TestResults should be populated and the
+// environment should not be nil.
+func (t *TestResults) Download(ctx context.Context) ([]TestResult, error) {
+	if !t.populated {
+		return []TestResult{}, errors.New("cannot download with populated test results")
+	}
 	if t.env == nil {
-		return errors.New("cannot append with a nil environment")
+		return []TestResult{}, errors.New("cannot download test results with a nil environment")
 	}
 
 	if t.ID == "" {
 		t.ID = t.Info.ID()
 	}
 
-	set := bson.M{}
-	for _, testName := range testNames {
-		set[bsonutil.GetDottedKeyName(testResultsArtifactKey, testResultsArtifactInfoTestsKey, testName)] = false
+	conf := &CedarConfig{}
+	conf.Setup(t.env)
+	if err := conf.Find(); err != nil {
+		return []TestResult{}, errors.Wrap(err, "problem getting application configuration")
 	}
-	updateResult, err := t.env.GetDB().Collection(testResultsCollection).UpdateOne(
+
+	bucket, err := t.Artifact.Type.Create(
 		ctx,
-		bson.M{"_id": t.ID},
-		bson.M{"$set": set},
+		t.env,
+		conf.Bucket.TestResultsBucket,
+		t.Artifact.Prefix,
+		string(pail.S3PermissionsPrivate),
+		false,
 	)
-	grip.DebugWhen(err == nil, message.Fields{
-		"collection":   testResultsCollection,
-		"id":           t.ID,
-		"updateResult": updateResult,
-		"test_names":   testNames,
-		"op":           "append test names to a test results record",
-	})
-	if err == nil && updateResult.MatchedCount == 0 {
-		err = errors.Errorf("could not find test results record with id %s in the database", t.ID)
+	if err != nil {
+		return []TestResult{}, errors.Wrap(err, "problem creating bucket")
+	}
+
+	iter, err := bucket.List(ctx, "")
+	if err != nil {
+		return []TestResult{}, errors.Wrap(err, "problem listing bucket items")
+	}
+
+	results := []TestResult{}
+	for iter.Next(ctx) {
+		r, err := iter.Item().Get(ctx)
+		if err != nil {
+			return []TestResult{}, errors.Wrap(err, "problem getting test result")
+		}
+		defer func() {
+			_ = r.Close()
+		}()
+
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return []TestResult{}, errors.Wrap(err, "problem reading test result data")
+		}
+		result := TestResult{}
+		if err = bson.Unmarshal(data, &result); err != nil {
```

Comment:
I think it might be a fun exercise to see what the bson unmarshal speed would be like vs writing DocumentMarshal and DocumentUnmarshaler functions in birch?